File: AndroidNative/games/src/main/java/com/chathil/games/transformer/GamesTransformer.kt
package com.chathil.games.transformer

import com.chathil.games.data.local.entity.GameEntity
import com.chathil.games.data.remote.response.GameResponse
import com.chathil.games.domain.model.GameModel

fun GameEntity.toDomainModel() = GameModel(
    id,
    slug,
    name,
    released,
    tba,
    backgroundImage,
    rating,
    ratingTop,
    ratings?.toDomainModels(),
    ratingsCount,
    reviewTextCount,
    added,
    metacritic,
    playtime,
    suggestionCount,
    reviewsCount,
    saturatedColor,
    dominantColor,
    platforms?.toDomainModels(),
    parentPlatforms?.toDomainModels(),
    genres?.toDomainModels(),
    stores?.toDomainModels(),
    clip,
    tags?.toDomainModels(),
    shortScreenshots?.toDomainModels()
)

fun List<GameEntity>.toDomainModels() =
    map { it.toDomainModel() }

fun GameResponse.toEntity() = GameEntity(
    id,
    name,
    slug,
    released,
    tba,
    backgroundImage,
    rating,
    ratingTop,
    ratings?.toEntities(),
    ratingsCount,
    reviewsTextCount,
    added,
    metacritic ?: 0,
    playtime,
    suggestionsCount,
    reviewsCount,
    saturatedColor,
    dominantColor,
    platforms?.toEntities(),
    parentPlatforms?.map { it.platform.toEntity() },
    genres?.toEntities(),
    stores?.toEntities(),
    clip?.clip,
    tags?.toEntities(),
    shortScreenshots?.toEntities()
)

fun List<GameResponse>.toEntities() = map { it.toEntity() }

File: iOSNative/Games/Sources/Games/Transformer/GamesTransformer.swift
//
//  File.swift
//  
//
//  Created by Abdul Chathil on 31/03/21.
//

import Foundation
import CoreData

extension GameResponse {
  func toEntity(context: NSManagedObjectContext) -> GameEntity {
    let entity = GameEntity(context: context)
    entity.id = id
    entity.slug = slug
    entity.name = name
    entity.released = released
    entity.tba = tba
    entity.backgroundImage = backgroundImage
    entity.rating = rating ?? 0.0
    entity.ratingTop = ratingTop ?? 0.0
    entity.ratingsArr = ratings.toEntitities()
    entity.ratingsCount = Int16(ratingsCount)
    entity.reviewTextCount = Int16(reviewsTextCount)
    entity.added = Int16(added)
    entity.metacritic = Int16(metacritic ?? 0)
    entity.playtime = Int16(playtime)
    entity.suggestionCount = Int16(suggestionsCount)
    entity.reviewsCount = Int16(reviewsCount)
    entity.saturatedColor = saturatedColor
    entity.dominantColor = dominantColor
    entity.platformsArr = platforms.toEntities()
    entity.parentPlatformsArr = parentPlatforms.toEntities()
    entity.genresArr = genres.toEntities()
    entity.storesArr = stores.toEntities()
    entity.clip = clip?.clip
    entity.tagsArr = tags.toEntities()
    entity.shortScreenshotsArr = shortScreenshots.toEntities()
    return entity
  }
}

extension Array where Element == GameResponse {
  func toEntities(context: NSManagedObjectContext) -> [GameEntity] {
    map { $0.toEntity(context: context) }
  }
}
